packages/realm-react/src/useEmailPasswordAuth.tsx

```diff
@@ -0,0 +1,190 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2023 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+import { useApp, useAuthResult } from "./AppProvider";
+import { AuthOperationName, AuthResult } from "./types";
+import { Realm, User, Credentials } from "realm";
+import { useAuthOperation } from "./useAuthOperation";
+
+/**
+ * Hook providing operations and corresponding state for authenticating with a
+ * Realm app with Email/Password.  It also contains operations related to
+ * Email/Password authentication, such as resetting password and confirming a user.
+ *
+ * The {@link AuthResult} states returned from this hook are "global" for all
+ * components under a given AppProvider, as only one operation can be in progress
+ * at a given time (i.e. we will store the states on the context). This means that,
+ * for example, multiple components can use the `useAuth` hook to access
+ * `result.pending` to render a spinner when login is in progress, without
+ * needing to pass that state around or store it somewhere global in their app
+ * code.
+ */
+interface UseEmailPasswordAuth {
+  /**
+   * Convenience function to login a user with an email and password - users
+   * could also call `logIn(Realm.Credentials.emailPassword(email, password)).
+   *
+   * @returns A `Realm.User` instance for the logged in user.
+   */
+  logIn(args: { email: string; password: string }): Promise<Realm.User | void>;
+
+  /**
+   * Register a new user. By default this will login the newly registered user
+   * after they have been successfully registered, unless the
+   * `loginAfterRegister` property is `false`.
+   *
+   * @returns A `Realm.User` instance for the logged in user if
+   * `loginAfterRegister` is not `false`, `void` otherwise.
+   */
+  register(args: {
+    email: string;
+    password: string;
+    // Defaults to true so that a newly registered user is immediately logged
+    // in.
+    loginAfterRegister?: boolean;
+  }): Promise<Realm.User | void>;
```

Comment:
Regarding `loginAfterRegister`, I think it could potentially create confusion having this default to `true` since the function has additional semantics than the name (it also logs in). For apps using Realm today, I'd assume that many users set up their apps so that after `register()` is successfully called, they also call `logIn()`. Even if this hook now provides an option to not have to explicitly perform the login step (which I think is great), the call will look the same (`register(credentials)`) but the semantics will differ.

Some alternatives are:
1. Provide `register()` without the `loginAfterRegister` and have it just be responsible for registration.
2. Provide `register()` with `loginAfterRegister` (or `logInAfterRegister`) default set to `false`.
3. Provide (a) `registerAndLogIn()` as well as (b) `register()` without the `loginAfterRegister`.

I think I'd prefer Alt. 3 (or some variation of that name). It also aligns with all the other convenience methods where we differentiate them via separate methods rather than an option param.

I'd be interested to hear your thoughts on this.

Reply:
Glad you brought this up, as I have been a bit torn on this as well.  It is not clear from the method that some side effect is happening by default.  I think I would actually lean towards option 3.  Then it is clear what is being attempted. 
One thought I'm having now though, is that if the LogIn fails, it may be unclear that the registration was successful.  There would be no roll-back in this case.  So maybe option 1 is just the better way to go to ensure the business logic can be setup exactly as the developer intends.

Reply:
Interesting point you mentioned with the "failed login but successful registration" scenario. I think Option 3 can still allow users to set up custom logic (providing `registerAndLogIn()` as well as `logIn()` and `register()`), while still catering to the use case of `registerAndLogIn()`. Since registering and logging in would throw different error messages, users could also differentiate between the errors in their `catch` clauses.

With our recent discussion regarding email confirmation, perhaps the more common case in production is to actually perform confirmation logic after registration before logging in.

Reply:
I'm going to pull the trigger on option 3.  The idea with doing both at the same time, was to make it even quicker to get started, but I don't think it's really worth the cost here.  We can always decide to add a combined version later, but we can't back out if we do it now.

Reply:
Good reasoning, go for it 👍